File: emisor/capas/aplicacion.py
"""Capa de Aplicacion del emisor (cajero automatico).

Solicita al usuario el mensaje a enviar, el algoritmo de integridad y la tasa de
error, y muestra la respuesta del servidor bancario.
"""

MENSAJE_DEFECTO = "RETIRO 100.00 TARJETA 23748"
TASA_DEFECTO = 0.01


def solicitar_mensaje(algoritmos_disponibles):
    """Pide texto, algoritmo y tasa de error. Devuelve (texto, algoritmo, tasa)."""
    texto = input(f"Mensaje a enviar [{MENSAJE_DEFECTO}]: ").strip() or MENSAJE_DEFECTO

    opciones = "/".join(algoritmos_disponibles)
    predeterminado = algoritmos_disponibles[0]
    algoritmo = input(
        f"Algoritmo ({opciones}) [{predeterminado}]: "
    ).strip().lower() or predeterminado

    entrada = input(f"Tasa de error por bit (0-1) [{TASA_DEFECTO}]: ").strip()
    tasa = float(entrada) if entrada else TASA_DEFECTO
    return texto, algoritmo, tasa


def mostrar_respuesta(respuesta):
    """Muestra la respuesta del banco al usuario del cajero."""
    if respuesta is None:
        print(">> Sin respuesta del servidor.")
        return
    estado = respuesta.get("estado", "?")
    mensaje = respuesta.get("mensaje", "")
    print(f">> [banco/{estado}] {mensaje}")

File: receptor/capas/aplicacion.js
// Capa de Aplicacion del receptor (servidor bancario).
// Autor: Fernando Hernandez (23645)
//
// Recorre la pila del receptor de abajo hacia arriba para cada paquete:
// Enlace (verificar/corregir) -> Presentacion (decodificar) -> mostrar el
// mensaje y armar la respuesta del banco para el cajero. Si se detectaron
// errores no corregibles, lo indica con un mensaje de error.

const enlace = require("./enlace");
const presentacion = require("./presentacion");

// Muestra el mensaje al receptor (o el error) y devuelve la respuesta que la
// capa de Transmision enviara de vuelta al cajero.
function mostrarMensaje(paquete) {
  const { algoritmo, trama, n_datos: nDatos } = paquete;

  // ENLACE: verificar integridad y corregir si el algoritmo lo permite.
  let r;
  try {
    r = enlace.verificarIntegridad(trama, algoritmo, nDatos);
  } catch (e) {
    console.log(`[banco] paquete rechazado: ${e.message}`);
    return { estado: "error", mensaje: e.message };
  }
  if (r.estado === "error") {
    const detalle = r.bloquesConError != null ? ` (${r.bloquesConError} bloques)` : "";
    console.log(`[banco] error no corregible${detalle}`);
    return { estado: "error", mensaje: "Trama con errores no corregibles" };
  }

  // PRESENTACION: ASCII binario -> texto.
  let texto;
  try {
    texto = presentacion.decodificarMensaje(r.bitsDatos);
  } catch (e) {
    console.log(`[banco] error al decodificar: ${e.message}`);
    return { estado: "error", mensaje: `Error al decodificar: ${e.message}` };
  }

  // APLICACION: mostrar el mensaje y responder al cajero.
  console.log(`[banco] ${r.estado}: "${texto}" (corregidos=${r.bloquesCorregidos ?? 0})`);
  return { estado: r.estado, mensaje: `Recibido: ${texto}` };
}

module.exports = {
  mostrarMensaje,
};
